### PR TITLE
refactor: add `BackendType.Primitive`

### PR DESCRIPTION
**Pull request by jaschdoc**
Depends on https://github.com/flix/flix/pull/6408
Closes https://github.com/flix/flix/issues/5917

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/jvm/BackendType.scala [hunk 1]
```diff
@@ -70,7 +64,7 @@ sealed trait BackendType extends VoidableType {
   /**
     * A string representing the erased type. This is used for parametrized class names.
     */
-  val toErasedString: String = this match {
+  def toErasedString: String = this match {
```
Comment (JonathanStarup) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/BackendType.scala:67]: what's the reason for this change?

File: main/src/ca/uwaterloo/flix/language/phase/jvm/BackendType.scala [hunk 2]
```diff
@@ -81,27 +75,54 @@ sealed trait BackendType extends VoidableType {
     case BackendType.Float64 => "Float64"
     case BackendType.Array(_) | BackendType.Reference(_) => "Obj"
   }
+
+  def is64BitWidth: Boolean = this match {
```
Comment (JonathanStarup) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/BackendType.scala:79]: documentation

File: main/src/ca/uwaterloo/flix/language/phase/jvm/BackendType.scala [hunk 3]
```diff
@@ -81,27 +75,54 @@ sealed trait BackendType extends VoidableType {
...
+    case BackendType.Int64 | BackendType.Float64 => true
+    case BackendType.Bool | BackendType.Char | BackendType.Int8 | BackendType.Int16 |
+         BackendType.Int32 | BackendType.Float32 | BackendType.Array(_) | BackendType.Reference(_) => false
+  }
+
+  /**
+    * Returns the Array fill type for the value of the type specified by `tpe`
```
Comment (JonathanStarup) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/BackendType.scala:86]: missing period at the end

File: main/src/ca/uwaterloo/flix/language/phase/jvm/BackendType.scala [hunk 4]
```diff
@@ -122,16 +143,69 @@ object BackendType {
...
+    case MonoType.Float64 => Float64
     case MonoType.Unit | MonoType.BigDecimal | MonoType.BigInt | MonoType.String | MonoType.Regex |
          MonoType.Array(_) | MonoType.Lazy(_) | MonoType.Ref(_) | MonoType.Tuple(_) |
          MonoType.Enum(_) | MonoType.Arrow(_, _) | MonoType.RecordEmpty | MonoType.RecordExtend(_, _, _) |
          MonoType.SchemaEmpty | MonoType.SchemaExtend(_, _, _) | MonoType.Native(_) |
          MonoType.Region => BackendObjType.JavaObject.toTpe
   }
+
+  /**
+    * Computes the `BackendType` based on the given `MonoType`.
```
Comment (JonathanStarup) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/BackendType.scala:160]: update documentation

File: main/src/ca/uwaterloo/flix/language/phase/jvm/BackendType.scala [hunk 5]
```diff
@@ -122,16 +143,69 @@ object BackendType {
...
+      // TODO: Ugly hack.
+      val fqn = clazz.getName.replace('.', '/')
+      BackendObjType.Native(JvmName.mk(fqn)).toTpe
+    case MonoType.Unit | MonoType.Lazy(_) | MonoType.Ref(_) |
+         MonoType.Tuple(_) | MonoType.Arrow(_, _) | MonoType.RecordEmpty |
+         MonoType.RecordExtend(_, _, _) | MonoType.Region | MonoType.Enum(_) |
+         MonoType.SchemaEmpty | MonoType.SchemaExtend(_, _, _) => BackendObjType.JavaObject.toTpe
+  }
+
+  sealed trait PrimitiveType {
```
Comment (JonathanStarup) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/BackendType.scala:187]: Every `PrimitiveType` is a `BackendType` so let `PrimitiveType` extend `BackendType` and then `Bool` will just extend `PrimitiveType` instead of both. Then you also don't need `toTpe`.

File: main/src/ca/uwaterloo/flix/language/phase/jvm/BackendType.scala [hunk 6]
```diff
@@ -17,48 +17,42 @@
 package ca.uwaterloo.flix.language.phase.jvm
 
 import ca.uwaterloo.flix.language.ast.MonoType
+import org.objectweb.asm.Opcodes._
```
Comment (JonathanStarup) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/BackendType.scala:20]: In my opinion `Opcodes._` shouldn't be imported, just write `Opcodes.AASTORE` fx at each spot

**Review by JonathanStarup — CHANGES_REQUESTED** [review]

File: main/src/ca/uwaterloo/flix/language/phase/jvm/BackendType.scala (hunk 1 above)
Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/phase/jvm/BackendType.scala:67]: I thought I had changed it to `val` so I wanted to revert it, but I see that's not the case.

File: main/src/ca/uwaterloo/flix/language/phase/jvm/BackendType.scala (hunk 4 above)
Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/phase/jvm/BackendType.scala:160]: Will be fixed in other pr

**Review by JonathanStarup — APPROVED** [review]

File: main/src/ca/uwaterloo/flix/language/phase/jvm/VoidableType.scala [hunk 7]
```diff
@@ -0,0 +1,23 @@
+package ca.uwaterloo.flix.language.phase.jvm
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/VoidableType.scala:1]: Copyright header.

File: main/src/ca/uwaterloo/flix/language/phase/jvm/VoidableType.scala [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/VoidableType.scala:3]: Extra space

**Review by magnus-madsen — APPROVED** [review]
